### ISSUE TITLE
Broadcast Receipts (data synchronization)

### ISSUE THREAD

**Issue by micylt** [dragonchain/dragonchain#65]
Implement node receipt functionality. Currently verification records are sent up the pipe to phase 2, 3, etc. but there is nothing in place for say a phase 2 to send back a signed verification recored receipt to phase 1. In other words, once any phase validates and signs a verification record, we want that node to send the signed verification record back to the phase that sent it.

Design:
Phase/node X sends record to phase Y; the next time phase X sends a record to phase Y, phase Y should send back all signed records it has on record for phase X, including records Y has for the X from nodes Z, A, B, etc. Ideally we would like all info being sent between nodes to be done on the same pipe. There should be a receipt cut-off point. If node Y has not sent back a receipt for a verification record by this point, we may want to have a configuration that says drop node Y (bad node) at that point.

**Comment by micylt** [comment]
@j0j0r0 and @Voxelot Please review and let me know if anything should be added/removed/modified. Thank you.

**Comment by j0j0r0** [comment]
Only thing that I would add is that the phase Y node should send back all signed records that it has for the calling node **that have not previously been returned to the calling node**

1. Node X sends record to Node Y
1. Node Y sends back all signed records it has for Node X, including records that it has signed, as well as any higher phases that it has received from other nodes

I expect that we will need to add a few fields to the datastore - I think in a new table (`broadcast_receipt`):

- broadcast attempts - count of attempted broadcasts to higher phased node(s) - for a broadcast to multiple phases (e.g. Phase 2), this should only be counted as 1 
- receipt confirmed (Boolean field to indicate that the signed record has been returned down to the calling node)

_I think that we need a new table because the broadcast attempts count to multiple nodes will need to be tracked individually for retry and to determine if a node is not answering. Potentially, we could store the attempts in a structured field, but I think that we'll want something more efficient for the query_

Any node that is not signing a Node X's record after some configured number of attempts can be marked as `bad` or `suspected_bad` and the network module can act to replace it.

If the next level has not been attained, then the processing module can continue to use nodes that are arranged by the network module to move forward.

Let me know if you have any questions.

**Comment by Voxelot** [comment]
I think eagerly returning upstream verification records could be a useful optimization, but it doesn't guarantee completeness. For the initial implementation, let's start with a simple cron-based pull engine. 

When a Phase 2+ node creates a local verification record or receives one from upstream, it should also insert a transmission record for each interested party with transmitted set to false. When a client to this node makes a request for any missing data it hasn't received, query the transmission table to check if there are any verification records that haven't been transmitted yet and return those. After successful transmission, we can either mark those records as "sent" or simply remove those transmission records from the table to cleanup.

To determine the interested parties for inclusion in the transfer table, find verification records with the same origin and block id with a lower phase than the current node, and create a transmission entry for each signatory.

We also need to handle data duplication. For example, the same phase 3 record may be replicated across several phase 2 nodes. When the origin node pulls the same phase 3 record from each phase 2 node, it would wastefully and redundantly increase consumption bandwidth. To avoid this, when a client node attempts to pull data, the current node should return a list of block verification ids that haven't been sent yet. In response to this, the client should take the list of returned block verification ids and filter out any that it's already received. The remaining list will be sent back to the phase 2 node as a request for the full data. The phase 2 node will return the requested data items, marking the corresponding entries in the transmission table as sent.

We should also stub or begin implementing some way to assign a fitness score to upstream nodes. When querying for available data, if the peer isn't returning expected data back in a timely manner we should de-prioritize it in the connection list.

Here's what a simple table schema may look like for keeping track of transmitted records:
```
BEGIN;
/* Don't drop tables unless you really want to */
--DROP TABLE block_transfers;
CREATE TABLE IF NOT EXISTS block_transfers (
    /* Original owner of the transaction data (Blockchain ID) */
    origin_id VARCHAR(256),
    /* The node to transmit this record to */
    transfer_to VARCHAR(256),
    verification_id UUID,
    sent BIT(1) DEFAULT 0
);
COMMIT;
BEGIN;
GRANT ALL ON block_transfer to blocky;
COMMIT;
```

And the sequence diagram for what I described above:
```
                                                                                                                     +---------+                                                                  +---------+
                                                                                                                     | Node_A  |                                                                  | Node_B  |
                                                                                                                     +---------+                                                                  +---------+
                                                                                                                          |                                                                            |
                                                                                                                          | query_available_data(Node_A.id)                                            |
                                                                                                                          |--------------------------------------------------------------------------->|
                                                                                                                          |                                                                            | ----------------------------------------------\
                                                                                                                          |                                                                            |-| select verification_id from block_transfers |
                                                                                                                          |                                                                            | | where transfer_to = Node_A.id               |
                                                                                                                          |                                                                            | |---------------------------------------------|
                                                                                                                          |                                                                            |
                                                                                                                          |                                         {"verification_ids": [1, 2, 3, 4]} |
                                                                                                                          |<---------------------------------------------------------------------------|
                                               -------------------------------------------------------------------------\ |                                                                            |
                                               | received, unreceived =                                                 |-|                                                                            |
                                               | split_items(verification_ids, id: verification_db.get(id) is not None) | |                                                                            |
                                               |------------------------------------------------------------------------| |                                                                            |
                                                                                                                          |                                                                            |
                                                                                                                          | transfer_data({"received":[1,2], "unreceived": [3,4]})                     |
                                                                                                                          |--------------------------------------------------------------------------->|
                                                                                                                          |                                                                            |
                                                                                                                          |      {"records": [{"verification_id":3, ...}, {"verification_id":4, ...}]} |
                                                                                                                          |<---------------------------------------------------------------------------|
                                                                                                                          |                                                                            | ------------------------------------------------------------------\
                                                                                                                          |                                                                            |-| mark records in block_transfers where                           |
                                                                                                                          |                                                                            | | verification_id = (1,2,3,4) and transfer_to = Node_A.id as sent |
                                                                                                                          |                                                                            | |-----------------------------------------------------------------|
------------------------------------------------------------------------------------------------------------------------\ |                                                                            |
| REPLICATION                                                                                                           |-|                                                                            |
| for each received verification record 'V':                                                                            | |                                                                            |
| select * from block_verifications where origin_id=V.origin_id and block_id=V.block_id and phase=V.phase-1             | |                                                                            |
|  for each lower phase verification found 'P':                                                                         | |                                                                            |
|  insert row into block_transfers with verification_id = V.verification_id, transfer_to = P.verification_info.signatory| |                                                                            |
|-----------------------------------------------------------------------------------------------------------------------| |                                                                            |
                                                                                                                          |                                                                            |
```

**Comment by j0j0r0** [comment]
Ok. 

Step 1 build the necessary tables, function(s) and queries to do the needed selection of applicable verification records. 

Step 2 decide when they will be called... 

If we can get requested VR or TX then we can move nimbly depending upon real measurements. 

But, the double call to avoid getting redundant data (which will only occur on a p2 node) looks to me like premature performance optimization in a different manner than that of the original design. 

We can see reuse of the queries for those cases where sync is necessary but I still think that direct response is both less complex and more efficient. 

A node knows which nodes were requested to sign, can best track its own data, and already has connection and is speaking with the nodes that will have the data. 

I re-propose that we (1) answer broadcast call with any records for the calling node (2) allow direct call for the same result in case the calling node has nothing to broadcast. 

The queries will likely be somewhat reused as well for the subscription data feed operation as Brandon mentioned.

**Comment by Voxelot** [comment]
I talked with @j0j0r0, and it seems like the best option is to have the broadcast call return a list of unsent verification ids. Then a second request would be made after the broadcast to retrieve unsent data items. This is because data duplication across the network may happen at phase 3 and phase 5. 

In addition to having the broadcast call return the list of ids, we should still have a background cron task to trigger this process in case there were no recent broadcasts. For each connected peer, timestamp the last time the verification records were retrieved and skip the cron task if the last sync was recent enough to avoid redundant calls.

**Comment by j0j0r0** [comment]
:+1: 

Yes - Broadcast call should return the list of unsent/unreturned verification record (VR) IDs.

A separate service GET would allow retrieval of non-retrieved verification records and would allow parameters to request or ignore specific IDs. Would also allow a HEAD request to get the list directly (w/o a broadcast call). An ignore parameter should mark the ignored VR as received on the called node.

**Comment by j0j0r0** [comment]
Merged #77 - closing ...
